### PR TITLE
Bobfile server

### LINKED ISSUE #39: Add PandA bob file server
First make an aiohttp project that serves `/<filename>.bob` returning "Hello <filename>" as the text

Then starting from https://github.com/PandABlocks/PandABlocks-client/tree/pvi_for_pythonsoftioc

- Integrate the aiohttp server into the startup of the ioc: https://github.com/PandABlocks/PandABlocks-client/blob/afe304fd9ad5b8f25ca4c7f7bd10319f164221b2/pandablocks/ioc/ioc.py#L109
- Make this code put the files in an internal dict: https://github.com/PandABlocks/PandABlocks-client/blob/pvi_for_pythonsoftioc/pandablocks/ioc/_pvi.py#L186-L198
- Serve those files on request from aiohttp

### PR DESCRIPTION
**Pull request by tizayi**
Resolves #39 .But needs tables branch of pvi to work.

### CODE REVIEW

File: tests/ioc/test_file_server.py
Comment (AlexanderWells-diamond) [inline tests/ioc/test_file_server.py:15]: It's worth turning all instances of `"TEST.bob"` into a constant. Same applies to the content of the file. This makes modifying/extending tests in the future a bit nicer.

File: tests/ioc/test_file_server.py
Comment (AlexanderWells-diamond) [inline tests/ioc/test_file_server.py:57]: Another constant it'd be good to extract to a named variable. Worth adding a comment to that this must match the name of the example .bob file.

File: tests/ioc/test_file_server.py
Comment (AlexanderWells-diamond) [inline tests/ioc/test_file_server.py:55]: I'm thinking that the host and port numbers should be defined as constants in `ioc.py` and imported here. I say that as I don't think we really care what the defaults are, so there's little point in testing them.

An alternative is to have another test that does hard-code these values, so that if they ever do change there's a test failure the coder has to investigate. The other tests can still use imported values.

File: tests/ioc/test_file_server.py
Comment (AlexanderWells-diamond) [inline tests/ioc/test_file_server.py:65]: What's the purpose of this write statement?

File: tests/ioc/test_file_server.py
Comment (AlexanderWells-diamond) [inline tests/ioc/test_file_server.py:57]: As we are only expecting this one item, I'd prefer slightly stricter `assert` statements, to confirm it is the one _and only_ item in the list

**Review by AlexanderWells-diamond — CHANGES_REQUESTED** [review]
This is a good start, but unfortunately there are a few bugs (some of which are mine!) remaining in the code. We'll need to clear them up before this is finished.

I think it might be worth waiting for a stable release of PVI with tables support before trying to merge this anywhere - the CI changes will all become obsolete as soon as that happens.

File: pandablocks/ioc/ioc.py
Comment (tizayi) [reply pandablocks/ioc/ioc.py:104]: I could also just leave the default 404 text.

Comment (AlexanderWells-diamond) [reply pandablocks/ioc/ioc.py:104]: I'd be happy with that

File: Pipfile
Comment (tizayi) [reply Pipfile:27]: I have removed the asynctest="*".

Using the aiohttp_client fixture from pytest-aiohttp resulted in the error:

`RuntimeError: Timeout context manager should be used inside a task`

After reading the source code for the fixture. I assumed this is an issue with the fixture creating a new event loop. So I created my own version of the fixture which uses the existing event loop.

Comment (tizayi) [reply Pipfile:27]: This code produces the error.


```python
import asyncio
import json
import os
from typing import Dict

import aiohttp
import pytest

from pandablocks.ioc.ioc import initialise_server
from tests.conftest import DummyServer

# Tests for the bobfile server
HOST = "0.0.0.0"
PORT = "8080"
PCAP1_FILE_NAME = "PCAP1.bob"
TEST_FILE_NAME = "TEST.bob"
TEST_FILE_CONTENTS = "<Bobfile>Contents</Bobfile>"
TEST_FILE_DIRECTORY = "test-bobfiles"

test_bob_file_dict: Dict[str, str] = {TEST_FILE_NAME: TEST_FILE_CONTENTS}


@pytest.mark.asyncio
async def test_get_available_files(aiohttp_client, event_loop):
    """Tests a request for the available files."""
    app = initialise_server(test_bob_file_dict)
    client = await aiohttp_client(
        app,
    )
    resp = await client.get("/")
    assert resp.status == 200
    text = await resp.text()
    assert text == f'["{TEST_FILE_NAME}"]'


@pytest.mark.asyncio
async def test_get_file(aiohttp_client, event_loop):
    """Tests a request for a single file."""
    app = initialise_server(test_bob_file_dict)
    client = await aiohttp_client(app)
    resp = await client.get(f"/{TEST_FILE_NAME}")
    assert resp.status == 200
    text = await resp.text()
    assert text == TEST_FILE_CONTENTS


@pytest.mark.asyncio
async def test_system_bobfile_creation(
    dummy_server_system: DummyServer, subprocess_ioc
):
    """A system test for both the bobfile creation and running the server."""
    loop = asyncio.get_event_loop()
    async with aiohttp.ClientSession(loop=loop) as session:

        async with session.get(f"http://{HOST}:{PORT}") as response:
            bob_file_list = json.loads(await response.text())
            assert PCAP1_FILE_NAME in bob_file_list

        async with session.get(f"http://{HOST}:{PORT}/{PCAP1_FILE_NAME}") as response:
            result = await response.text()
            dir_path = os.path.dirname(os.path.realpath(__file__))
            with open(dir_path + f"/{TEST_FILE_DIRECTORY}/{PCAP1_FILE_NAME}", "r") as f:
                assert result == f.read()

```

Comment (AlexanderWells-diamond) [reply Pipfile:27]: So it looks like this is a known issue - infact I raised this issue 5 months ago on the pytest-aiohttp plugin github: https://github.com/aio-libs/pytest-aiohttp/issues/43

As it appears that that this plugin is no longer maintained, you can ignore any comments I've made about switching to it!

Comment (AlexanderWells-diamond) [reply Pipfile:27]: Having said that... I've just reviewed another MR, which does make use of pytest-aiohttp, apparently successfully. Perhaps you can look at these tests and work out what the missing piece is? 
https://github.com/dls-controls/coniql/blob/7c312b9a1ff1cb34c110c887d91794e804c4b782/tests/test_aiohttp.py

File: pandablocks/ioc/_pvi.py
Comment (AlexanderWells-diamond) [inline pandablocks/ioc/_pvi.py:185]: Just spotted this TODO - it's no longer correct, so please delete it

File: pandablocks/ioc/ioc.py
Comment (AlexanderWells-diamond) [inline pandablocks/ioc/ioc.py:99]: There's a couple of typos here, and I can't tell what word `please` is supposed to be.

File: pandablocks/ioc/ioc.py
Comment (AlexanderWells-diamond) [inline pandablocks/ioc/ioc.py:68]: This appears to be unused?

**Review by AlexanderWells-diamond — APPROVED** [review]
Looking good. Just a couple of very minor nitpicks left. 

It would be good to get @coretl  and/or @GDYendell to comment on various PVI-related aspects, but apart from that I think we're good.

File: pandablocks/ioc/_pvi.py
Comment (coretl) [reply pandablocks/ioc/_pvi.py]: https://github.com/epics-containers/pvi/issues/23

File: pandablocks/ioc/ioc.py
Comment (coretl) [reply pandablocks/ioc/ioc.py]: This will do for now

**Review by coretl — APPROVED** [review]

**Pull request closed without merging**
